Learn C# Tutorial

BillWagner opened the pull request:
Learn the structure of a C# program, and  explore the string class.

[Internal review link](https://review.docs.microsoft.com/en-us/dotnet/csharp/quick-starts/?branch=pr-en-us-2306)

BillWagner committed:
First draft of new tutorial

Learn the structure of a C# program, and  explore the string class.

proofread and tighten language.

respond to feedback

2nd draft

Move more code earlier.
Tighten vocabulary discussions.

Update outline

start on the next draft.

add section on searching strings

fix build errors

finish the string tutorial.

reorder: ready for proofread.

proofread

update to use new syntax

Update Tutorial to use the new engine

typo in the title

BillWagner commented:
This is a work in progress.  Tasks required before merging:

- [ ] Polish the samples
- [ ] Test runs for time required (aiming for 30 minutes to perform all steps)
- [ ] Add cross references and "Additional Resources"
- [ ]

cartermp commented on docs/csharp/getting-started/interactive-lessons/hello-world.md:
```diff
@@ -0,0 +1,274 @@
+sample. These elements form the structure of C# programs. Let's examine them
+before you move on to modifying this program. Rather than working line-by-line,
+let's examine the building blocks from the inside out. This program, like all
+programs, is made up of *statements*, *methods*, *classes*, and *namespaces*.
```
:+1:, this is statement:

> This program, like all programs, is made up of *statements*, *methods*, *classes*, and *namespaces*.

Is only true for C#.  Other languages don't have have statements.  We should clarify that this is for "all _C#_ programs".

BillWagner committed:
build test, try samples

BillWagner committed:
2nd draft of the numbers tutorial.

mairaw commented:
@BillWagner I see that you removed the WIP label. Is this ready for review now?

BillWagner committed:
proofread the second tutorial and create the third

Branching and looping is ready for initial review.

mairaw commented:
@BillWagner this PR is bringing other PRs in. Can you fix that so commits/files not concerning the interactive tutorial are not included in this review?

mairaw commented on docs/csharp/quick-starts/hello-world.yml:
```diff
@@ -0,0 +1,171 @@
+- title: Run your first c# program
+  durationInMinutes: 2
+  content: |
+    Run the following code in the interactive window (type the following line in the window on the right and click the 'Run' button):
```
Instead of putting in parentheses, I suggest putting in a new sentence.
For example:
Run the following code in the interactive window. To do that, type (or maybe copy?) the following line...

mairaw commented on docs/csharp/quick-starts/hello-world.yml:
````diff
@@ -0,0 +1,171 @@
+    Console.WriteLine("Hello World!");
+    ```
+
+    Congratulations! You've run your first C# program. Now, replace the two lines above with these:
````
before you move, should you explain what they've done in the first code? 

Also, this sentence seems odd "Now, replace the two lines above with these:". To start, there was only one line.

mairaw commented on docs/csharp/quick-starts/hello-world.yml:
````diff
@@ -0,0 +1,171 @@
+    Run the following code in the interactive window (type the following line in the window on the right and click the 'Run' button):
+
+    ```csharp
+    Console.WriteLine("Hello World!");
````
question about the interactive window: it doesn't seem to enforce the need of ending the instruction with ; on the last line of code. By design?

BillWagner replied:
/cc @LadyNaggaga would have the answer to that.

mairaw commented on docs/csharp/quick-starts/hello-world.yml:
```diff
@@ -0,0 +1,171 @@
+  durationInMinutes: 3
+  content: |
+    Your first program printed the `string` "Hello World!" on
+    a screen. The rest of this lesson explores working with
```
a screen -> the screen?

mairaw commented on docs/csharp/quick-starts/hello-world.yml:
```diff
@@ -0,0 +1,171 @@
+    a screen. The rest of this lesson explores working with
+    the `string` type, which represents text in C#.
+
+    Your first program is limited to printing one message. You can write more
```
is limited -> was limited

mairaw commented on docs/csharp/quick-starts/hello-world.yml:
```diff
@@ -0,0 +1,171 @@
+
+    Your first program is limited to printing one message. You can write more
+    useful programs by using **variables**. A **variable** is a symbol you can
+    use run the same code with different values. Let's try it.
```
use run -> use to run

end with exclamation point?

mairaw commented on docs/csharp/quick-starts/hello-world.yml:
````diff
@@ -0,0 +1,171 @@
+    Console.WriteLine(firstName);
+    ```
+
+    Those lines declare a variable, `name` and assigns it a value, "Bill". Then, it
````
Maybe break down what each line does? The first line declares ... Then, the second line...

mairaw commented on docs/csharp/quick-starts/hello-world.yml:
```diff
@@ -0,0 +1,171 @@
+    prints out the name.
+
+    You can assign different values to any variable you declare. You can change
+    the name to one of your friends. Add these two lines in the interactive window above:
```
two lines? I'm seeing four 😄 

also the interactive windows is on the right not above

JRAlexander replied:
Remove any mention of direction. On smaller screens, the interactive windows are below.

mairaw replied:
Good point @JRAlexander!

mairaw commented on docs/csharp/quick-starts/hello-world.yml:
```diff
@@ -0,0 +1,171 @@
+    Notice that the same line of code prints two different messages, based on
+    the value stored in the `firstName` variable.
+
+    You may have also noticed the "Hello" was missing in the last two messages. Let's fix
```
noticed -> noticed that?

mairaw commented on docs/csharp/quick-starts/hello-world.yml:
```diff
@@ -0,0 +1,171 @@
+    the value stored in the `firstName` variable.
+
+    You may have also noticed the "Hello" was missing in the last two messages. Let's fix
+    that now.  Modify the last line to the following:
```
you're promising to fix that in the last two messages but then you only fix for one

mairaw commented on docs/csharp/quick-starts/hello-world.yml:
```diff
@@ -0,0 +1,171 @@
+  content: |
+    Your last edit was our first look at what you can do with strings. Let's explore more.
+
+    The `+`, when applied to strings, creates a new string with the text from the first followed by the text from the second. This is referred to as *string concatenation*. You can combine any number of strings, including string variables.  Try this:
```
I'd extend the introduction. Perhaps 
The `+` is called an operator in C#. When applied...

mairaw commented on docs/csharp/quick-starts/hello-world.yml:
````diff
@@ -0,0 +1,171 @@
+    Console.WriteLine("My friends are " + firstName + " and " + aFriend);
+    ```
+
+    As you explore more with strings, you can find the length of a string using `Length`.  Add the following code at the bottom of the interactive window:
````
Do you need to say what Length is?

mairaw commented on docs/csharp/quick-starts/hello-world.yml:
```diff
@@ -0,0 +1,171 @@
+- title: Do more with strings
+  durationInMinutes: 4
+  content: |
+    Another useful method is Trim, and related methods TrimStart and TrimEnd. These methods remove leading and trailing spaces:
```
you haven't introduced methods yet so it shouldn't be another useful method. I'd also add that methods can perform actions or something like that and then show Trim and co.

JRAlexander replied:
Another idea might be to link here: https://docs.microsoft.com/en-us/dotnet/csharp/tour-of-csharp/classes-and-objects#methods

mairaw commented on docs/csharp/quick-starts/hello-world.yml:
````diff
@@ -0,0 +1,171 @@
+  content: |
+    Another useful method is Trim, and related methods TrimStart and TrimEnd. These methods remove leading and trailing spaces:
+
+    ```csharp
````
another question for interactive: can we get a clear button?

BillWagner replied:
/cc @LadyNaggaga @EisenbergEffect

JRAlexander replied:
And maybe a Paste option on the context menu?

mairaw replied:
Good idea @JRAlexander. Just tried on my phone and without a paste button, it's almost impossible to follow this.

mairaw commented on docs/csharp/quick-starts/hello-world.yml:
```diff
@@ -0,0 +1,171 @@
+
+    This sample reinforces a couple important concepts for working with strings. The methods that manipulate strings return new string objects rather than making modifications in place. You can see that each call to any of the Trim methods returns a new string, but does not change the original message.
+
+    Other methods perform more general edits on a string. You've probably used a search and replace command in an editor or word processor. The Replace method substitutes a replacement string for all occurences of a sub-string in a string. Try it for yourself. Add this code above:
```
typo: occurences

I don't think substring needs hyphen

the explanation of Replace is a bit confusing. Maybe introduce the concept of parameters here? Could say something like "The Replace method searches for all occurrences of the first substring and replaces them with the second substring in a string"?

JRAlexander replied:
Or link to here: https://docs.microsoft.com/en-us/dotnet/api/system.string.replace?view=netcore-2.0#System_String_Replace_System_String_System_String_  😄

mairaw commented on docs/csharp/quick-starts/hello-world.yml:
````diff
@@ -0,0 +1,171 @@
+    Console.WriteLine(message);
+    ```
+
+    Two other useful methods make a string ALL CAPS or all lower case. Try this code above:
````
Try the following code?

mairaw commented on docs/csharp/quick-starts/hello-world.yml:
```diff
@@ -0,0 +1,171 @@
+  durationInMinutes: 4
+  content: |
+    The other part of a *search and replace* operation is to find text in a
+    sring. You can use the the `Contains` method for searching. It
```
typo: sring

mairaw commented on docs/csharp/quick-starts/hello-world.yml:
```diff
@@ -0,0 +1,171 @@
+  content: |
+    The other part of a *search and replace* operation is to find text in a
+    sring. You can use the the `Contains` method for searching. It
+    tells you if a method contains a sub-string inside it. Try the following
```
if a method -> if a string

mairaw commented on docs/csharp/quick-starts/hello-world.yml:
````diff
@@ -0,0 +1,171 @@
+    code to explore `Contains`:
+
+    ```csharp
+    string message = "You say goodbye, and I say hello";
````
nit: can we add periods to these sentences? 😄

mairaw commented on docs/csharp/quick-starts/hello-world.yml:
```diff
@@ -0,0 +1,171 @@
+    ***Challenge***
+
+    There are two similar methods, `StartsWith` and `EndsWith` that also search for sub-strings in a string. These find a sub-string at the beginning or the
+    end of the string. Try to modify the sample above to use `StartsWith` and
```
above -> previous

mairaw commented on docs/csharp/quick-starts/hello-world.yml:
```diff
@@ -0,0 +1,171 @@
+
+    There are two similar methods, `StartsWith` and `EndsWith` that also search for sub-strings in a string. These find a sub-string at the beginning or the
+    end of the string. Try to modify the sample above to use `StartsWith` and
+    `EndsWith` instead of `Contains`.
```
maybe continue the explanation. Otherwise, I'd expect that beginners would simply change the Contain methods with StartsWith or EndsWith

mairaw commented on docs/csharp/quick-starts/hello-world.yml:
````diff
@@ -0,0 +1,171 @@
+    Console.WriteLine(message.EndsWith("hello"));
+    Console.WriteLine(message.EndsWith("goodbye"));
+    ```
+- content: |
````
another feedback about interactive: should Next steps, Next tutorial be sentence case?

BillWagner replied:
/cc @EisenbergEffect

mairaw commented on docs/csharp/quick-starts/hello-world.yml:
```diff
@@ -0,0 +1,171 @@
+    You've completed the "Hello C#" tutorial.
+
+    You can learn more about working with the `string` type in the
+    [C# Programming Guide](../programming-guide/index.md) topic on [strings](../programming-guide/strings/index.md)
```
missing punctuation.

mairaw commented on docs/csharp/quick-starts/index.md:
```diff
@@ -0,0 +1,52 @@
+---
+title: Quick Starts - C# Guide | Microsoft Docs
```
modify breadcrumb file to add quick starts there

mairaw reviewed (changes requested):
This is going to be awesome @BillWagner. I've reviewed the first tutorial for now.

BillWagner commented on docs/csharp/quick-starts/index.md:
```diff
@@ -0,0 +1,52 @@
+---
+# Quick starts
+
+Welcome to the C# Quick starts. These start with interactive lessons
```
not yet. I removed it.

BillWagner committed:
respond to feedback on first tutorial.

JRAlexander commented on docs/csharp/quick-starts/working-with-numbers.yml:
````diff
@@ -0,0 +1,264 @@
+    ```
+
+    You may have noticed an interesting behavior for integers. Integer
+    division always produces an integer result, even when the mathematic
````
Should this be mathematical instead of mathematic?

JRAlexander commented on docs/csharp/quick-starts/working-with-numbers.yml:
```diff
@@ -0,0 +1,264 @@
+    There are other numeric types with different limits and precision that you
+    would use when the `int` type does not meet your needs. Let's explore those next.
+
+- title: Working with the double type.
```
Remove the period.

JRAlexander commented on docs/csharp/quick-starts/working-with-numbers.yml:
````diff
@@ -0,0 +1,264 @@
+    ```
+
+    Notice that the answer includes the decimal portion of the quotient. Try a slightly
+    more complicted expression with doubles:
````
complicted -> complicated.

JRAlexander commented on docs/csharp/quick-starts/working-with-numbers.yml:
````diff
@@ -0,0 +1,264 @@
+    ```
+
+    These values are printed out to represent scientific notation. The number to
+    the left of the `E` is the signifant. The number to the right is the exponent,
````
signifant -> significant.

JRAlexander commented on docs/csharp/quick-starts/working-with-numbers.yml:
```diff
@@ -0,0 +1,264 @@
+    Another part represents where to place the binary point (analogous to a decimal point) relative to
+    those significant digits.
+
+        Just like decimal numbers in math, doubles in C# can have rounding error. Try this code:
```
error -> errors.

JRAlexander commented on docs/csharp/quick-starts/working-with-numbers.yml:
```diff
@@ -0,0 +1,264 @@
+    The `M` suffix on the numbers is how you indicate that a constant should use the
+    `decimal` type.
+
+    Notice that the math done using the decimal type has more digits to the right
```
Consider removing "done".

JRAlexander reviewed (changes requested):
Overall, great work, @BillWagner! Just a few items.

BillWagner committed:
proof read 3rd tutorial

And add links to other C# Guide content

BillWagner committed:
Remove links to upcoming tutorials.

mairaw commented on docs/csharp/quick-starts/hello-world.yml:
```diff
@@ -0,0 +1,176 @@
+  durationInMinutes: 4
+  content: |
+    The other part of a *search and replace* operation is to find text in a
+    string. You can use the the `Contains` method for searching. It
```
the the

link to the API?

mairaw commented on docs/csharp/quick-starts/hello-world.yml:
````diff
@@ -0,0 +1,176 @@
+    code to explore `Contains`:
+
+    ```csharp
+    string message = "You say goodbye, and I say hello.";
````
either this sentence doesn't have punctuation or you change your challenge to search for a period in the end of the string. Might be easier to do the former.

mairaw commented on docs/csharp/quick-starts/hello-world.yml:
````diff
@@ -0,0 +1,176 @@
+    ```
+
+    The `Contains` method returns a **boolean** value which tells you if the
+    string you sought was found. A **boolean** stores either a `true` or a
````
wondering if sought is a hard term for ESL speakers. Maybe replace "you sought" with "you were searching for"?

mairaw commented on docs/csharp/quick-starts/hello-world.yml:
```diff
@@ -0,0 +1,176 @@
+
+    ***Challenge***
+
+    There are two similar methods, `StartsWith` and `EndsWith` that also search for sub-strings in a string. These find a sub-string at the beginning or the
```
substring, no hyphen

mairaw commented on docs/csharp/quick-starts/index.md:
```diff
@@ -0,0 +1,47 @@
+keywords: C#, Get Started, Lessons, Interactive
+author: billwagner
+ms.author: wiwagn
+ms.date: 06/01/2017
```
you might want to update this date?

mairaw reviewed (commented):
Argh, again I haven't gone past the point of the first tutorial. I'll continue tonight.

BillWagner committed:
Respond to feedback.

BillWagner committed:
fix yaml parsing error.

mairaw commented on docs/csharp/quick-starts/hello-world.yml:
```diff
@@ -0,0 +1,180 @@
+- title: Do more with strings
+  durationInMinutes: 4
+  content: |
+    You've been using a **method**, <xref:System.Console.WriteLine>, to print messages. A **method** is a block of code that implements some action. It has a name, so you can access it. 
```
use ?displayProperty=fullName here

mairaw commented on docs/csharp/quick-starts/numbers-in-csharp.yml:
```diff
@@ -0,0 +1,271 @@
+    
+    You know that `0.3` repeating is not exactly the same as `1/3`.
+
+        ***Challenge***
```
this is rendering inside a code block, probably not by design.

mairaw commented on docs/csharp/quick-starts/numbers-in-csharp.yml:
```diff
@@ -0,0 +1,271 @@
+    You've seen the basic numeric types in C#: integers and doubles.  There is one
+    other type to learn: the `decimal` type. The `decimal` type has a smaller
+    range, but greater precision than `double`. The term **fixed point** means
+    that the decimal point (or binary point) does not move. Let's take a look:
```
nit: does not -> doesn't

mairaw commented on docs/csharp/quick-starts/numbers-in-csharp.yml:
````diff
@@ -0,0 +1,271 @@
+    ```csharp
+    decimal min = decimal.MinValue;
+    decimal max = decimal.MaxValue;
+    Console.WriteLine("the range of the decimal type is " + min + " to " + max);
````
the -> The

mairaw commented on docs/csharp/quick-starts/numbers-in-csharp.yml:
```diff
@@ -0,0 +1,271 @@
+- title: Complete challenge
+  durationInMinutes: 1
+  content: |
+    Did you come up with something like this?
```
I'd separate the complete challenge from next steps as you did in the hello world tutorial. Otherwise, the window with the code disappears and I can't double check if I got my code right.

mairaw commented on docs/csharp/quick-starts/numbers-in-csharp.yml:
````diff
@@ -0,0 +1,271 @@
+
+    ```csharp
+    double radius = 2.50;
+    double area = Math.PI * radius * radius;
````
should you also show the alternative using Math.Pow or too early?

mairaw commented on docs/csharp/quick-starts/numbers-in-csharp.yml:
````diff
@@ -0,0 +1,271 @@
+    double area = Math.PI * radius * radius;
+    Console.WriteLine(area);
+    ```
+    You've completed the "Numbers in C#" tutorial. You can learn more about numbers in C# in this topics:
````
this topics -> the following topics

mairaw commented on docs/csharp/quick-starts/branches-and-loops.yml:
```diff
@@ -0,0 +1,241 @@
+- title: Make decisions using the if statement
+  durationInMinutes: 2
+  content: |
+    Run the following code in the interactive window. To do that, type the following code block in the interactive window and click the *Run* button:
```
should we give a quick intro into booleans and if statement on this page? seems to go too straight to the point perhaps? could be before you introduced the first code.

BillWagner replied:
I want to 'hook' beginners with code, so I started there.

I did add a paragraph of text after running the first sample. Let me know what you think.

mairaw commented on docs/csharp/quick-starts/branches-and-loops.yml:
```diff
@@ -0,0 +1,241 @@
+
+    > [!IMPORTANT]
+    > The indentation under the `if` and `else` statements is for human readers.
+    > The C# language does not treat indentation or whitespace as significant. 
```
nit: does not -> doesn't

mairaw commented on docs/csharp/quick-starts/branches-and-loops.yml:
````diff
@@ -0,0 +1,241 @@
+    }
+    ```
+
+- title: use loops to repeat operations
````
use -> Use

mairaw commented on docs/csharp/quick-starts/branches-and-loops.yml:
```diff
@@ -0,0 +1,241 @@
+    > false as you execute the code. Otherwise, you create an
+    > **infinite loop** where your program never ends. Let's
+    > not demonstrate that, because the engine that runs your
+    > code will timeout and you'll see no output from your program.
```
timeout -> time out

mairaw commented on docs/csharp/quick-starts/branches-and-loops.yml:
```diff
@@ -0,0 +1,241 @@
+    `for` loop continues to execute as long as the value of counter is less than 10.
+
+    The final part is the **for iterator**: `counter++` specifies how to modify the loop
+    variable after executing the block following the `for` statement. Here, it specirifies
```
typo: specirifies

mairaw commented on docs/csharp/quick-starts/branches-and-loops.yml:
```diff
@@ -0,0 +1,241 @@
+
+    - Change the initializer to start at a different value.
+    - Change the condition to stop at a different value.
+    - Change the iterator to add a value other than 1 to `counter`.
```
isn't it too early to ask this? most people will use counter++ or counter-- in their for loops. they have just been introduced to counter++. will they know they should self assign a value here like counter += 2?

BillWagner replied:
Good point.

mairaw commented on docs/csharp/quick-starts/branches-and-loops.yml:
```diff
@@ -0,0 +1,241 @@
+  durationInMinutes: 12
+  content: |
+    Now that you've seen the `if` statement and the looping
+    constructs in the C# language. See if you can write C# to
```
C# language. See -> C# language, see

write C# -> write C# code.

mairaw commented on docs/csharp/quick-starts/branches-and-loops.yml:
```diff
@@ -0,0 +1,241 @@
+  content: |
+    Now that you've seen the `if` statement and the looping
+    constructs in the C# language. See if you can write C# to
+    find the some of all integers 1 through 20 that are divisible
```
some -> sum?

BillWagner replied:
ROFL

mairaw commented on docs/csharp/quick-starts/branches-and-loops.yml:
```diff
@@ -0,0 +1,241 @@
+    find the some of all integers 1 through 20 that are divisible
+    by 3.  Here are a few hints:
+    
+    - The `%` operator will give you the remainder of a division operation.
```
will give -> gives

mairaw commented on docs/csharp/quick-starts/branches-and-loops.yml:
```diff
@@ -0,0 +1,241 @@
+    
+    Let's move on to one last loop statement.
+
+- title: Work with the `for` loop.
```
remove period and perhaps the inline code too?

![image](https://user-images.githubusercontent.com/12971179/30581701-e720838c-9cd5-11e7-86f9-a1dafafff781.png)

mairaw commented on docs/csharp/quick-starts/numbers-in-csharp.yml:
```diff
@@ -0,0 +1,271 @@
+    int a = 18;
+    int b = 6;
+    int c = a + b;
+    Console.WriteLine(c);
```
I'm wondering if we should write in the text what the customer should see so they can make sure they're doing it right.

BillWagner replied:
I kept the numbers small so readers should be able to check their work quickly. How strongly do you feel about this?

mairaw replied:
I don't feel strongly about this right now. We can see how this goes and change if needed.

mairaw reviewed (commented):
Hopefully my final comments!

mairaw commented on docs/csharp/quick-starts/branches-and-loops.yml:
```diff
@@ -0,0 +1,241 @@
+    When you're done, let's move on to write some code yourself to
+    use what you've learned.
+
+- title: Challenge yourself
```
should we be consistent on titles? Hello world uses Complete challenge

mairaw commented on docs/csharp/quick-starts/branches-and-loops.yml:
```diff
@@ -0,0 +1,241 @@
+- title: Complete challenge
+  durationInMinutes: 1
+  content: |
+    Did you come up with something like this?
```
this window also erases the result. I suggest you break down result and next steps in two like you did with hello world

BillWagner committed:
More feedback.

BillWagner committed:
feedback from team review

Configure ms.technology for all quick starts to be
devlang-csharp-interactive.
Update .NET home page and C# Guide home page to point to the
interactive tutorials.

BillWagner committed:
Describe booleans, and update metadata.

BillWagner committed:
add a tip about compiler messages

BillWagner committed:
update timings of tutorials, update tip text

mairaw commented on docs/csharp/quick-starts/index.md:
```diff
@@ -0,0 +1,46 @@
+within minutes. These first lessons assume no prior knowledge of
+programming or the C# language.
+
+## [Hello world](hello-world)
```
the link works but OPS doesn't like this. Usually we need to add the extension too. Not sure how this is done with quick-starts. Do we need to add yml to the link so OPS doesn't give us a warning? /cc @eisenbergeffect

mairaw commented on docs/csharp/quick-starts/numbers-in-csharp.yml:
```diff
@@ -0,0 +1,285 @@
+### YamlMime:YamlDocument
+documentType: Tutorial
```
can we have this recognized by OPS build? /CC @eisenbergeffect

mairaw reviewed (approved):
I'm approving the PR but there are a few things that need to be fixed before merging. Other comments are just nits or weird things with OPS.

BillWagner committed:
final review.
